### PR TITLE
Fixes #35948 - Support for incremental syncable exports

### PR DESCRIPTION
**Pull request by parthaa**
#### What are the changes introduced in this pull request?
This commit adds the ability to do incremental syncable exports. 
We allowed only complete before for syncable and now you can do incremental. 

```
$ hammer content-export incremental repository --id=1 --format=syncable
[....................................................................................................................................................................] [100%]
Generated /var/lib/pulp/exports/Default_Organization/Export-SYNCABLE-interesting-1/12.0/2023-01-13T22-59-47-00-00/metadata.json
```

The files are exported in a syncable format but only contain rpms + metadata that changed since the last time you exported. 
cd into `/var/lib/pulp/exports/Default_Organization/Export-SYNCABLE-interesting-1/12.0/2023-01-13T22-59-47-00-00/` and you should see a directory structure that looks like the following. (In my repo I added a new duck rpm and it had gotten regenerated correctly and so you'd only notice 1 rpm)
```
$ cd /var/lib/pulp/exports/Default_Organization/Export-SYNCABLE-interesting-1/12.0/2023-01-13T22-59-47-00-00/
$ find custom
custom
custom/media
custom/media/interesting
custom/media/interesting/Packages
custom/media/interesting/Packages/d
custom/media/interesting/Packages/d/duck-0.8-1.noarch.rpm
custom/media/interesting/.treeinfo
custom/media/interesting/repodata
custom/media/interesting/repodata/68b9c4b0c0163b650160d76f691fd34cb1a2f18664b0136b965583ab3ad58640-primary.xml.gz
custom/media/interesting/repodata/e4ec28d7d0482b6db237b9a0a756fe704e28ca9bc73405b290e75d9723613848-filelists.xml.gz
custom/media/interesting/repodata/fcdaaa6cd37b259a599f01fc38f740ac50edb85de25227acb286b55797ec25bd-other.xml.gz
custom/media/interesting/repodata/3ae1b862f0d906e0f0f268bbccb16346c631d3dd6ed656fbf5dcba29c6f1f8bb-updateinfo.xml.gz
custom/media/interesting/repodata/540ff6c3a42ee01ab43f1d87b7b09d8d8787c5a505baa2a8764db31adc76c3d7-modules.yaml
custom/media/interesting/repodata/4eb81ce725d3a48773d1497d93acf9aed1cf2aa11da24ba228095b6edc372cba-comps.xml
custom/media/interesting/repodata/repomd.xml
```

You can then use this to sync or import into a downstream server via hammer content-import

##### Other Changes  #####
- Adds a setting for pulpcore_export_format which you can specify syncable/importable by default (so don't need to use --format)
- DRY'es up the content-export and content-incremental-export controllers (they had mostly duplicated descriptions and content) and puts the common code in exports-controller (which gets included in both controllers.)

#### Considerations taken when implementing this change?
So this is useful for a customer who wants to export repos in a syncable format.
Pulpcore PR => https://github.com/pulp/pulpcore/pull/3511 now adds the ability to do incremental syncable exports
Katello is taking advantage of this feature to implement incremental `syncable` exports.

All three content-export options are available to be extracted in syncable format.
- `hammer content-export incremental version --format=syncable ....`
- `hammer content-export incremental library --format=syncable ....`
- `hammer content-export incremental repository --format=syncable ....`

The export rpm files are generated as hardlinks to the actual file on disk. This there by avoids the 2X space we may need for regular importable exports.  

#### What are the testing steps for this pull request?
- On your dev env you may need to run `sudo setfacl --recursive --modify u:vagrant:rwX,d:u:vagrant:rwX /var/lib/pulp/exports` so that listing files can be generate when hammer is running as a vagrant user.
- On your katello instance `Administer -> Settings -> Content`. 
- set both `Default Custom Repository download policy` and `Default Red Hat Repository download policy` to immediate
- Create an org
- Add/Enable + sync few repos. Custom repos will be easier to check but RH should work just the same. 
- Create a content view with these repos an publish.
- Do a syncable complete export. One of
  - `hammer content-export complete version --format=syncable ....`
  - `hammer content-export complete library --format=syncable ....`
  - `hammer content-export complete repository --format=syncable ....`

For example:
```
$ hammer content-export complete repository --organization=org --product=prod --name=repo
[....................................................................................................................................................................] [100%]
Generated /var/lib/pulp/exports/org/Export-SYNCABLE-repo-19/1.0/2023-01-14T20-26-50-00-00/metadata.json
$ find /var/lib/pulp/exports/org/Export-SYNCABLE-repo-19/1.0/2023-01-14T20-26-50-00-00/ |grep rpm|wc
     27      27    3643
# 27 rpms in this list
```

- Upload an rpm to the custom repo so that a new version is created
 - Publish the content-view again
 Try any of the 3 hammer commands listed here
- `hammer content-export incremental version --format=syncable ....`
- `hammer content-export incremental library --format=syncable ....`
- `hammer content-export incremental repository --format=syncable ....`
```
$ hammer content-export incremental repository --organization=org --product=prod --name=repo --format=syncable
[....................................................................................................................................................................] [100%]
Generated /var/lib/pulp/exports/org/Export-SYNCABLE-repo-19/2.0/2023-01-15T04-23-20-00-00/metadata.json

$ find /var/lib/pulp/exports/org/Export-SYNCABLE-repo-19/2.0/2023-01-15T04-23-20-00-00/|grep rpm|wc
      1       1     130
```

- Make sure you only see the added rpm and new repo metadata

### CODE REVIEW

**Comment by killroy1971** [comment]
This is close.  You can't merge incremental repository metadata, and if this is added to the existing exported repository metadata, Foreman will dump the older metadata. This would open errata from registered systems. In the above example, the only package metadata would be for the new package. 

Instead of incremental repository metadata, we will need a full repository metadata export. I'm not sure if there is an incremental repository metadata use case.  That way, the system administrator can backup the existing metadata into a tarball, copy in the new metadata, and validate the updated repository content and errata with a downstream disconnected Foreman/Satellite server.

**Comment by parthaa** [comment]
[test katello]

File: app/controllers/katello/concerns/api/v2/exports_controller.rb [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  module Concerns
+    module Api::V2::ExportsController
+      extend ActiveSupport::Concern
+
+      class_methods do
+        def export_format_param
+          export_format_description = N_("Export formats."\
+                               "Choose syncable if the exported content needs to be in a yum format. "\
+                               "This option is only available for %{syncable_repos} repositories. "\
+                               "Choose importable if the disconnected importing server uses the same version "\
```
Comment (ianballou) [inline app/controllers/katello/concerns/api/v2/exports_controller.rb:11]: Does the importing server need to be disconnected?

```suggestion
                               "Choose importable if the importing server uses the same version "\
```

**Review by ianballou — COMMENTED** [review]
A few comments:

File: app/controllers/katello/concerns/api/v2/exports_controller.rb [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        def destination_server_param
+          param :destination_server, String, :desc => N_("Destination Server name"), :required => false
+        end
+
+        def from_history_id_param
+          param :from_history_id, :number, :desc => N_("Export history identifier used for incremental export. "\
+                                             "If not provided the most recent export history will be used."), :required => false
+        end
+
+        def export_library_description(incremental: false)
```
[the remark this answers is not in the document]
Comment (chris1984) [reply app/controllers/katello/concerns/api/v2/exports_controller.rb:54]: Is this what you are referring to? https://stackoverflow.com/questions/35821499/combine-params-required-with-apipie-param-group

File: app/controllers/katello/api/v2/content_export_incrementals_controller.rb [hunk 3]
```diff
@@ -8,69 +9,29 @@ class Api::V2::ContentExportIncrementalsController < Api::V2::ApiController
...
-                          chunk_size: params[:chunk_size_gb],
-                          from_history: @history,
-                          fail_on_missing_content: ::Foreman::Cast.to_bool(params[:fail_on_missing_content]))
-
-      respond_for_async :resource => tasks
+      export_content_view_version
     end
 
     api :POST, "/content_export_incrementals/library", N_("Performs an incremental-export of the repositories in library.")
-    param :organization_id, :number, :desc => N_("Organization identifier"), :required => true
```
[the remark this answers is not in the document]
Comment (parthaa) [reply app/controllers/katello/api/v2/content_export_incrementals_controller.rb:31]: I kinda addressed your suggestion i think. Let me know if it works for you

File: app/controllers/katello/concerns/api/v2/exports_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (parthaa) [reply app/controllers/katello/concerns/api/v2/exports_controller.rb]: @ianballou  Tried the param groups approach. Check out the controllers and let me know if you prefer this ?

**Comment by ianballou** [comment]
The code is looking good to me now! @parthaa do you think you need another test due to the changes you made?

**Comment by parthaa** [comment]
> The code is looking good to me now! @parthaa do you think you need another test due to the changes you made?

Sure now that the new pulp core for 3.21.5 is available . I recommend updating pulpcore and trying this (instead of manually patching)

**Comment by ianballou** [comment]
In testing, I'm hitting: `ActionController::RoutingError (No route matches [POST] "/api/content_export_incrementals/version")`

The apipie docs are fine, but it seems the routing is messed up. Restarting Rails doesn't help.

The same thing happens for the other incremental export types.

```
[vagrant@centos7-hammer-devel ~]$  hammer content-export incremental version --format=syncable --id 6
Could not export the content view version:
  404 Not Found
```

**Comment by parthaa** [comment]
> In testing, I'm hitting: `ActionController::RoutingError (No route matches [POST] "/api/content_export_incrementals/version")`
> 
> The apipie docs are fine, but it seems the routing is messed up. Restarting Rails doesn't help.
> 
> The same thing happens for the other incremental export types.
> 
> ```
> [vagrant@centos7-hammer-devel ~]$  hammer content-export incremental version --format=syncable --id 6
> Could not export the content view version:
>   404 Not Found
> ```
> 
> Edit, in apipie I see: `POST /api/content_export_incrementals/version `
> 
> Looks like the path is missing the Katello part for some reason! And hammer seems to be dropping the Katello as well. It seems the routing is being placed under the Foreman API namespace for some reason.
> 
> Edit edit: It looks like full exports are affected too.

Glad you tested :) I am getting the same. Will update

**Comment by ianballou** [comment]
I think I may have found a bug -- I create a CV with 3 yum repos and some filters. I then made a complete export of the content view version.  Then, I removed all the filters and tried making an incremental export. Hammer returned with a 500. In the logs I saw the error:

```
The exported Content View Version 'Syncable Incremental Export 2.0' cannot be incrementally updated from version '1.0'. Please do a full export.
```

I'm guessing the issue is related to the fact that we generate new repositories when we're filtering, and when we're not, we "redirect" to copies of the library repositories. Perhaps the exporting logic cannot be incremental across repositories like that?

CVV 1.0 Zoo repo href: /pulp/api/v3/repositories/rpm/rpm/08cf534a-3af2-4d88-aa45-bb296e0a1063/versions/1/
CVV 2.0 Zoo repo href: /pulp/api/v3/repositories/rpm/rpm/fa0550c1-2a66-43b1-a140-1755df3c0a66/versions/2/

I'm guessing that's the issue: different Pulp repositories under the hood.

I suppose you could cover this by adding an error condition for incrementally exporting between CVVs that have filters and not, but that sounds kinda like a funny user experience.

Also, the error didn't bubble up to Hammer. I only saw a 500.

```
[vagrant@centos7-hammer-devel ~]$  hammer content-export incremental version --format=syncable --id 6
Could not export the content view version:
  500 Internal Server Error
```

**Comment by ianballou** [comment]
I found another issue I think -- I cannot delete my repositories anymore after creating syncable exports. I get the following error: 
```
ActiveRecord::DeleteRestrictionError

Cannot delete record because of dependent library_instances_inverse
```

And I checked it out and found this:

```
[1] pry(main)> ::Katello::Repository.find(29).library_instances_inverse
=> [#<Katello::Repository:0x00005560bcfce848
  id: 33,
  pulp_id: "1-Export-Library-SYNCABLE-v2_0-6bfd7e20-0fc4-4b48-8318-5c1eae397f46",
  library_instance_id: 29,
  content_view_version_id: 7,
  relative_path: "Default_Organization/content_views/Export-Library-SYNCABLE/2.0/custom/Buttermilk_Pancakes/repo",
  environment_id: nil,
  saved_checksum_type: nil,
  distribution_version: nil,
  distribution_arch: nil,
  distribution_bootable: nil,
  distribution_family: nil,
  distribution_variant: nil,
  container_repository_name: nil,
  root_id: 9,
  remote_href: nil,
  publication_href: "/pulp/api/v3/publications/rpm/rpm/fac27628-4ed6-48ff-b361-c68b172333e3/",
  version_href: "/pulp/api/v3/repositories/rpm/rpm/d46d3e7f-0e54-401f-ab4f-c70f8dbd3172/versions/1/",
  last_contents_changed: Mon, 06 Feb 2023 19:53:08.686389000 UTC +00:00,
  last_applicability_regen: Thu, 01 Jan 1970 00:00:00.000000000 UTC +00:00,
  last_indexed: Mon, 06 Feb 2023 19:53:11.016295000 UTC +00:00>,
 #<Katello::Repository:0x00005560bcfce5f0
  id: 34,
  pulp_id: "1-Export-Library-SYNCABLE-Library-6bfd7e20-0fc4-4b48-8318-5c1eae397f46",
  library_instance_id: 29,
  content_view_version_id: 7,
  relative_path: "Default_Organization/Library/Export-Library-SYNCABLE/custom/Buttermilk_Pancakes/repo",
  environment_id: 1,
[2] pry(main)> 
```

**Comment by parthaa** [comment]
> I suppose you could cover this by adding an error condition for incrementally exporting between CVVs that have filters and not, but that sounds kinda like a funny user experience.

I thought we did that error check. will figure out

**Comment by parthaa** [comment]
> I thought we did that error check. will figure out

What would a better message be ?

**Comment by parthaa** [comment]
@ianballou @jeremylenz @wbclark how does this sound? Need some wordsmithing here :)
```
The exported Content View Version 'Syncable Incremental Export 2.0' cannot be incrementally updated from version '1.0'. You many not be able to export due one of the following reasons
1. You are trying to export between a filtered and non-filtered content view versions. 
2. You are trying to export between an incrementally updated content view version and regular version  
3. You are trying to incrementally export from importable to syncable, or vice-versa.

Please do a full export.
```

**Comment by ianballou** [comment]
Looks pretty good! Just some tiny adjustments:

```
The exported Content View Version 'Syncable Incremental Export 2.0' cannot be incrementally updated from version '1.0'. You many not be able to export due one of the following reasons:
1. You are trying to export between a filtered and a non-filtered content view version. 
2. You are trying to export between an incrementally updated content view version and regular version  
3. You are trying to incrementally export from importable to syncable, or vice versa.

Please do a full export.
```

**Comment by jeremylenz** [comment]
Can it tell me which of the reasons it is?

remember to add the "to" :)

```
 You many not be able to export due to one of the following reasons:
```

**Comment by ianballou** [comment]
> Can it tell me which of the reasons it is?

We could tell if it's reason 2 if one of the CVVs has a y version and the other does not.  Reason 1 we would have to infer from it not being reason 2 since we don't keep a filter history.  We should likely be able to tell as well if it's reason 3.

We'd need to be careful because if we miss a potential reason, the result could be confusing for the user.

**Comment by ianballou** [comment]
@parthaa there are some Katello test failures

**Comment by ianballou** [comment]
I'm seeing the incremental flag get set incorrectly on the `metadata.json`.

So I did the following:

```
[vagrant@centos7-hammer-devel ~]$  hammer -r content-export incremental version --format=syncable --id 10
[.....................................................................................................................................] [100%]
Unable to access/write to '/var/lib/pulp/exports/Default_Organization/Syncable_Incremental_Export/3.0/2023-02-13T20-21-36-00-00'. Generated '/home/vagrant/metadata-7.json' instead. This file is necessary to perform an import.
```

That was between my version 1 and version 3 content view versions. Both are all filtered.

If I look in `metadata-7.json`:

```json
{
    "base_path": "/var/lib/pulp/exports",
    "content_view": {
        "description": "",
        "generated_for": "none",
        "label": "Syncable_Incremental_Export",
        "name": "Syncable Incremental Export"
    },
    "content_view_version": {
        "description": "",
        "major": 3,
        "minor": 0
    },
    "format": "syncable",
    "gpg_keys": {},
    "incremental": false,
    "organization": "Default Organization",
    "products": {
        "Buttermilk_Pancakes": {
            "cp_id": "903496789511",
            "description": null,
            "gpg_key": {},
            "label": "Buttermilk_Pancakes",
            "name": "Buttermilk Pancakes",
            "redhat": false
        },
        "Red_Hat_Enterprise_Linux_for_x86_64": {
            "cp_id": "479",
            "description": null,
            "gpg_key": {},
            "label": "Red_Hat_Enterprise_Linux_for_x86_64",
            "name": "Red Hat Enterprise Linux for x86_64",
            "redhat": true
        }
    },
    "repositories": {
        "Red_Hat_Enterprise_Linux_9_for_x86_64_-_BaseOS_RPMs_9-109803": {
            "arch": "noarch",
            "checksum_type": null,
            "content": {
                "id": "11049",
                "label": "rhel-9-for-x86_64-baseos-rpms",
                "url": "/content/dist/rhel9/9/x86_64/baseos/os"
            },
            "content_type": "yum",
            "description": null,
            "download_policy": "immediate",
            "gpg_key": {},
            "label": "Red_Hat_Enterprise_Linux_9_for_x86_64_-_BaseOS_RPMs_9",
            "major": 9,
            "minor": "9",
            "mirroring_policy": "mirror_complete",
            "name": "Red Hat Enterprise Linux 9 for x86_64 - BaseOS RPMs 9",
            "os_versions": [],
            "product": {
                "label": "Red_Hat_Enterprise_Linux_for_x86_64"
            },
            "redhat": true,
            "unprotected": false
        },
        "Red_Hat_Satellite_Client_6_for_RHEL_9_x86_64_RPMs-95082": {
            "arch": "noarch",
            "checksum_type": null,
            "content": {
                "id": "12626",
                "label": "satellite-client-6-for-rhel-9-x86_64-rpms",
                "url": "/content/dist/layered/rhel9/x86_64/sat-client/6/os"
            },
            "content_type": "yum",
            "description": null,
            "download_policy": "immediate",
            "gpg_key": {},
            "label": "Red_Hat_Satellite_Client_6_for_RHEL_9_x86_64_RPMs",
            "major": null,
            "minor": null,
            "mirroring_policy": "mirror_complete",
            "name": "Red Hat Satellite Client 6 for RHEL 9 x86_64 RPMs",
            "os_versions": [],
            "product": {
                "label": "Red_Hat_Enterprise_Linux_for_x86_64"
            },
            "redhat": true,
            "unprotected": false
        },
        "Zoo-88788": {
            "arch": "noarch",
            "checksum_type": null,
            "content": {
                "id": "1672780978853",
                "label": "Default_Organization_Buttermilk_Pancakes_Zoo",
                "url": "/custom/Buttermilk_Pancakes/Zoo"
            },
            "content_type": "yum",
            "description": null,
            "download_policy": "immediate",
            "gpg_key": {},
            "label": "Zoo",
            "major": null,
            "minor": null,
            "mirroring_policy": "additive",
            "name": "Zoo",
            "os_versions": [],
            "product": {
                "label": "Buttermilk_Pancakes"
            },
            "redhat": false,
            "unprotected": true
        }
    }
}
```

Why is there ""incremental": false" ?

**Comment by ianballou** [comment]
I tried doing this for library and repository and for some reason I never see `incremental: true`.  How should I test that the export is actually incremental?

**Comment by ianballou** [comment]
Also, we should make sure that the fix goes in for the error-propagation issues -- that's really gonna confuse folks since we really rely on errors as part of the UX.

**Comment by ianballou** [comment]
Mechanics-wise I think I'm seeing what I should be seeing. The full export contains all the goodies in a "repo format".  The incremental export does as well, but only contains the content that is the difference.

**Comment by parthaa** [comment]
> I tried doing this for library and repository and for some reason I never see `incremental: true`. How should I test that the export is actually incremental?

should be fixed now. Nice catch. Was not sending the from_content_view_version to the metadata. Shouldn't affect the workflow in the case of syncable.

**Comment by ianballou** [comment]
The incremental flag in the JSON seems to be correct now!

**Review by ianballou — APPROVED** [review]

**Pull request merged**